### PR TITLE
Allow modes `z` and `Z` for docker volumes

### LINKED ISSUE #18989: `z` option is mishandled when adding custom `docker_volumes`
my job conf includes:

```
<param id="docker_volumes">$galaxy_root:ro,$tool_directory:ro,$job_directory:ro,$working_directory:z,$default_file_path:z</param>
```

there are five entries. Generated tool scripts look like:

```
-v /home/user/arbeit/galaxy/galaxy:/home/user/arbeit/galaxy/galaxy:ro \                                                                 
-v /home/user/arbeit/emc/infrastructure/files/emc-tools:/home/user/arbeit/emc/infrastructure/files/emc-tools:ro \                       
-v /home/user/arbeit/galaxy/galaxy/database/jobs_directory/000/499:/home/user/arbeit/galaxy/galaxy/database/jobs_directory/000/499:ro \ 
-v /home/user/arbeit/galaxy/galaxy/database/objects:z:rw \ 
--cpus ${GALAXY_SLOTS:-1}
```

note that there are now four `-v` instead of 5, (missing working_directory, I think. `outputs_to_working_directory` is false.

additionally, and more importantly, `:rw` has been added to the end of the mount, and the mount hasn't been reduplicated like usual. I'm guessing this is because there's some code that checks specifically for `:ro` and `:rw`, but doesn't yet consider `:z` or `:Z` (needed when running under selinux.) and assumes those are mount paths inside the container.

### PR DESCRIPTION
**Pull request by bernt-matthias**
Fixes https://github.com/galaxyproject/galaxy/issues/18989

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

### CODE REVIEW

**Comment by bernt-matthias** [comment]
> special handling around singularity

Will see what I can do.

**Comment by bernt-matthias** [comment]
Is it correct that one can comma separate bind options, e.g. `rw,z`?

https://docs.docker.com/engine/storage/volumes/#choose-the--v-or---mount-flag: "The third field is optional, and is a comma-separated list of options, such as ro. These options are discussed below." (unfortunately I can't find the "below"...

I was also asking myself if it might be an alternative to hardcode the corresponding binds as additional docker options?

**Comment by hexylena** [comment]
> Is it correct that one can comma separate bind options, e.g. rw,z?

testing locally i can confirm, `rw,z` and `z,rw` are both valid

I also cannot find the below in their docs :thinking: 

> I was also asking myself if it might be an alternative to hardcode the corresponding binds as additional docker options?

I don't believe `$default_file_path` and `$working_directory` are available in the additional options field? if they were it would indeed be an option to put it there once you've set `<param id="docker_volumes"></param>` to be intentionally empty.

**Comment by hexylena** [comment]
I ran this patch today and noted the following:

```
            Error: z: duplicate mount destination
```

it seems the mounts are being interpreted as a literal `z`

```
-v $dir/galaxy/database/jobs_directory/000/503/working:z -v $dir/galaxy/database/objects:z 
```

rather than being reduplicated like happens for rw/ro

**Comment by bernt-matthias** [comment]
Guess we **could** simplify the code by dropping https://github.com/galaxyproject/galaxy/blob/d11d8930a34cc7ef5aa7a72388c7c57348a14c14/lib/galaxy/tool_util/deps/container_volumes.py#L52 ..

File: lib/galaxy/tool_util/deps/container_volumes.py [hunk 1]
```diff
@@ -7,7 +7,7 @@
 
 
 class ContainerVolume(metaclass=ABCMeta):
-    valid_modes = frozenset({"ro", "rw"})
+    valid_modes = frozenset({"ro", "rw", "z"})
```
Comment (hexylena) [inline lib/galaxy/tool_util/deps/container_volumes.py]:
```suggestion
    valid_modes = frozenset({"ro", "rw", "z", "Z"})
```

**Comment by hexylena** [comment]
In my testing it's still generating single element volumes, I'm not sure why. My reading of the code suggests that shouldn't be done anymore but, I'll need to look at it further.

**Comment by hexylena** [comment]
with 2d7cd6f, `-v /home/user/arbeit/galaxy/galaxy/database/jobs_directory/000/505:ro:rw -v /home/user/arbeit/galaxy/galaxy/database/objects:z:rw`

**Comment by bernt-matthias** [comment]
My last commit should (I hope) make it testable. Could you add you test case? Then we may be able to understand what is going on.

**Comment by hexylena** [comment]
No, that's got it completely! 

```
/usr/bin/podman run -e "GALAXY_SLOTS=$GALAXY_SLOTS" -e "GALAXY_MEMORY_MB=$GALAXY_MEMORY_MB" -e "GALAXY_MEMORY_MB_PER_SLOT=$GALAXY_MEMORY_MB_PER_SLOT" -e "HOME=$HOME" -e "_GALAXY_JOB_HOME_DIR=$_GALAXY_JOB_HOME_DIR" -e "_GALAXY_JOB_TMP_DIR=$_GALAXY_JOB_TMP_DIR" -e "TMPDIR=$TMPDIR" -e "TMP=$TMP" -e "TEMP=$TEMP" --name a21df026676e41b38646c0fe14511ef4 \
-v /home/user/arbeit/galaxy/galaxy:/home/user/arbeit/galaxy/galaxy:ro \
-v /home/user/arbeit/emc/infrastructure/files/emc-tools:/home/user/arbeit/emc/infrastructure/files/emc-tools:ro \
-v /home/user/arbeit/galaxy/galaxy/database/jobs_directory/000/506:/home/user/arbeit/galaxy/galaxy/database/jobs_directory/000/506:ro \
-v /home/user/arbeit/galaxy/galaxy/database/jobs_directory/000/506/working:/home/user/arbeit/galaxy/galaxy/database/jobs_directory/000/506/working:z \
-v /home/user/arbeit/galaxy/galaxy/database/objects:/home/user/arbeit/galaxy/galaxy/database/objects:z \
--cpus ${GALAXY_SLOTS:-1} -w /home/user/arbeit/galaxy/galaxy/database/jobs_directory/000/506/working --rm --userns
```

yeah that looks completely correct now.

**Review by hexylena — APPROVED** [review]
Thanks for fixing this @bernt-matthias

**Pull request merged**